src/cargo/ops/cargo_rustc/job_queue.rs

```diff
@@ -259,8 +254,11 @@ impl<'a> Dependency for (&'a PackageId, Stage) {
             // dependencies (making them available to us).
             Stage::BuildCustomBuild => {
                 let mut base = vec![(id, Stage::Start)];
-                base.extend(deps.filter(|&(_, dep)| dep.is_build())
-                                .map(|(id, _)| (id, Stage::Libraries)));
+                base.extend(deps
+                    .filter(|id| pkg.dependencies().iter()
+                        .find(|d| d.name() == id.name() && d.is_build())
+                        .is_some())
+                    .map(|id| (id, Stage::Libraries)));
```

Comment:
This formatting is a bit confusing because the alignment of `.filter` with `.extend` normally indicates that the `filter` is happening on the return value of `extend`, perhaps this could be formatted as:

```rust
                base.extend(deps.filter(|id| {
                    pkg.dependencies().iter().find(|d| {
                        d.name() == id.name() && d.is_build()
                    }).is_some()
                }).map(|id| (id, Stage::Libraries)));

```